### PR TITLE
Integration test/test repository resource

### PR DESCRIPTION
**Pull request by F4lka**
### Checklist
- [x] Server: I added multiple integration tests (Spring) related to the features

### Motivation and Context

This PR is part of the [initiative](https://confluence.ase.in.tum.de/display/ArTEMiS/Test+Coverage)   to increase the test coverage for the REST controllers.

### Description

added several test cases to cover all REST calls and at least 80% line coverage

### Screenshot
Screenshot from 2020-05-05 14-04-55
Screenshot from 2020-05-05 14-04-45

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void shouldCreateFile() throws Exception {
+        programmingExerciseRepository.save(exercise);
+        LinkedMultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+        assertThat(Files.exists(Paths.get(testRepo.localRepoFile + "/newFile"))).isFalse();
+        params.add("file", "newFile");
+        request.postWithoutResponseBody(testRepoBaseUrl + exercise.getId() + "/file", HttpStatus.OK, params);
+        assertThat(Files.exists(Paths.get(testRepo.localRepoFile + "/newFile"))).isTrue();
```
Comment (sleiss) [inline src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:107]: Maybe use `File.separator` instead of `/`.

Comment (sleiss) [inline src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:107]: Please check that `newFile` is a file and not a folder.

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void shouldCreateFolder() throws Exception {
+        programmingExerciseRepository.save(exercise);
+        LinkedMultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+        assertThat(Files.exists(Paths.get(testRepo.localRepoFile + "/newFolder"))).isFalse();
+        params.add("folder", "newFolder");
+        request.postWithoutResponseBody(testRepoBaseUrl + exercise.getId() + "/folder", HttpStatus.OK, params);
+        assertThat(Files.exists(Paths.get(testRepo.localRepoFile + "/newFolder"))).isTrue();
```
Comment (sleiss) [inline src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:118]: Please check that `newFolder` is a folder and not a file.

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void shouldRenameFile() throws Exception {
```
Comment (sleiss) [inline src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:123]: Maybe also test renaming a folder.

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        request.postWithoutLocation(testRepoBaseUrl + exercise.getId() + "/commit", null, HttpStatus.OK, null);
+        var receivedStatusAfterCommit = request.get(testRepoBaseUrl + exercise.getId(), HttpStatus.OK, RepositoryStatusDTO.class);
+        assertThat(receivedStatusAfterCommit.repositoryStatus.toString()).isEqualTo("CLEAN");
+    }
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void shouldPullChanges() throws Exception {
+        programmingExerciseRepository.save(exercise);
+        request.get(testRepoBaseUrl + exercise.getId() + "/pull", HttpStatus.OK, Void.class);
```
Comment (sleiss) [inline src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:161]: There is no assert here, is this correct?

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java (hunk 3 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:131]: i added a test case for renaming a folder

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java (hunk 1 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java]: i added a check for that

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java (hunk 2 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java]: i added a check for that

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java (hunk 1 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java]: I tested it with the File.separator, it worked but I think the readability decreases with the usage of the File.separator. Therefore, i would keep the current implementation.

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void shouldResetToLastCommit() throws Exception {
+        programmingExerciseRepository.save(exercise);
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:189]: 1) first create some "local" changes without committing them (in the best case, use the online code editor to add and modify files). If this is not possible, then directly manipulate the local git
2) Then create a merge conflict by changing the same files in the originGit
TODO: check if the Artemis server detects the conflict
3) Then invoke reset
4) Then check, that all local changes are removed and there is no conflict any more

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        database.resetDatabase();
+        reset(gitService);
+        LocalRepository.resetLocalRepo(testRepo);
+    }
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void shouldGetFiles() throws Exception {
+        programmingExerciseRepository.save(exercise);
+        var files = request.getMap(testRepoBaseUrl + exercise.getId() + "/files", HttpStatus.OK, String.class, FileType.class);
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:92]: check that the files actually exists, some of them might get filtered by the way?

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java (hunk 4 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java]: i added an assert to this method

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java (hunk 6 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java]: i added a check, which checks if the files actually exist

File: src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        this.localRepoFile = Files.createTempDirectory(localRepoFileName).toFile();
+        this.localGit = Git.init().setDirectory(localRepoFile).call();
+
+        this.originRepoFile = Files.createTempDirectory(originRepoFileName).toFile();
+        this.originGit = Git.init().setDirectory(originRepoFile).call();
+
+        this.localGit.remoteAdd().setName("origin").setUri(new URIish(String.valueOf(this.originRepoFile))).call();
+    }
+
+    public static void resetLocalRepo(LocalRepository localRepo) throws IOException {
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java:36]: please convert this into a non static method that is invoked on itself.

Comment (krusche) [reply src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java:36]:
```suggestion
    public void resetLocalRepo() throws IOException {
```

Comment (krusche) [reply src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java:36]: localRepo --> this

File: src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+        if (localRepo.originRepoFile != null && localRepo.originRepoFile.exists()) {
+            FileUtils.deleteDirectory(localRepo.originRepoFile);
+        }
+        if (localRepo.originGit != null) {
+            localRepo.originGit.close();
+        }
+    }
+
+    public static List<RevCommit> getAllCommits(Git gitRepo) throws Exception {
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java:52]: I would also prefer non static methods here. You could distinguish between 

public static List<RevCommit> getAllLocalCommits()

and 

public static List<RevCommit> getAllOriginCommits(Git gitRepo)

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        var receivedStatusAfterReset = request.get(testRepoBaseUrl + programmingExercise.getId(), HttpStatus.OK, RepositoryStatusDTO.class);
+        assertThat(receivedStatusAfterReset.repositoryStatus.toString()).isEqualTo("CLEAN");
+    }
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void shouldGetStatus() throws Exception {
+        programmingExerciseRepository.save(programmingExercise);
+        var receivedStatus = request.get(testRepoBaseUrl + programmingExercise.getId(), HttpStatus.OK, RepositoryStatusDTO.class);
+        assertThat(receivedStatus).isNotNull();
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java:287]: can you be a bit more specific here when testing for the correct status?

File: src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java (hunk 7 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java]: i adapted the method

File: src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java (hunk 8 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java]: I adapted the method and divided it into two methods -> getAllLocalCommits & getAllOriginCommits

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java (hunk 9 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java]: i adapted the testcase

File: src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java (hunk 5 above)
Comment (F4lka) [reply src/test/java/de/tum/in/www1/artemis/TestRepositoryResourceIntegrationTest.java]: i adapted the method

**Pull request merged**